File: src/tensorflow2.x-python-tutorial/linear_regression_keras.py
import numpy as np
import keras
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

X = np.array([3.3, 4.4, 5.5, 6.71, 6.93, 4.168, 9.779, 6.182, 7.59, 2.167,
                         7.042, 10.791, 5.313, 7.997, 5.654, 9.27, 3.1])
Y = np.array([1.7, 2.76, 2.09, 3.19, 1.694, 1.573, 3.366, 2.596, 2.53, 1.221,
                         2.827, 3.465, 1.65, 2.904, 2.42, 2.94, 1.3])

tf.config.run_functions_eagerly(True)

input = keras.layers.Input(shape=(1), name="data_in")

output = Dense(1)(input)
model = keras.Model(inputs=input, outputs=output)

# 选定loss函数和优化器
model.compile(loss='mse', optimizer=keras.optimizers.SGD(0.005), metrics='acc')

model.summary()

model.fit(X, Y, epochs=100)

W, b = model.layers[1].get_weights()

print("weight: ", W, "bias", b)

File: src/TensorFlowNET.Examples/BasicModels/LinearRegressionKeras.cs
/*****************************************************************************
   Copyright 2020 Haiping Chen. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
******************************************************************************/

using Tensorflow.NumPy;
using static Tensorflow.Binding;
using static Tensorflow.KerasApi;

namespace TensorFlowNET.Examples
{
    public class LinearRegressionKeras : SciSharpExample, IExample
    {
        NDArray train_X, train_Y;

        public ExampleConfig InitConfig()
            => Config = new ExampleConfig
            {
                Name = "Linear Regression (Keras)",
                Enabled = true,
                IsImportingGraph = false
            };

        public bool Run()
        {
            tf.enable_eager_execution();

            PrepareData();

            BuildModel();

            return true;
        }

        public override void BuildModel()
        {
            var inputs = keras.Input(shape: 1);
            var outputs = layers.Dense(1).Apply(inputs);
            var model = keras.Model(inputs, outputs);

            model.summary();
            model.compile(loss: keras.losses.MeanSquaredError(),
                optimizer: keras.optimizers.SGD(0.005f),
                metrics: new[] { "acc" });
            model.fit(train_X, train_Y, epochs: 100);

            var weights = model.TrainableVariables;
            print($"weight: {weights[0].numpy()}, bias: {weights[1].numpy()}");
        }

        public override void PrepareData()
        {
            train_X = np.array(3.3f, 4.4f, 5.5f, 6.71f, 6.93f, 4.168f, 9.779f, 6.182f, 
                    7.59f, 2.167f, 7.042f, 10.791f, 5.313f, 7.997f, 5.654f, 9.27f, 3.1f);

            train_Y = np.array(1.7f, 2.76f, 2.09f, 3.19f, 1.694f, 1.573f, 3.366f, 
                    2.596f, 2.53f, 1.221f, 2.827f, 3.465f, 1.65f, 2.904f, 2.42f, 2.94f, 1.3f);
        }
    }
}
